apps/notification-producer/src/producers/trade/TradeNotificationProducer.ts

```diff
@@ -80,26 +76,79 @@ export class TradeNotificationProducer implements Runnable {
     // Get last block
     const client = viemClients[chainId];
     const lastBlock = await client.getBlock();
-    const toBlock = lastBlock.number;
 
-    // Get trade events from block to last block
-    const fromBlock = stateRegistry?.state
+    const toBlockFinal = lastBlock.number;
+
+    // Get starting block
+    let fromBlock = stateRegistry?.state
       ? BigInt(stateRegistry.state.lastBlock) + 1n
-      : toBlock;
+      : toBlockFinal;
 
-    const numberOfBlocksToIndex = toBlock - fromBlock + 1n;
+    const totalBlocksToIndex = toBlockFinal - fromBlock + 1n;
 
     // Print debug message
-    if (numberOfBlocksToIndex < 1n) {
+    if (totalBlocksToIndex < 1n) {
       // We are up to date. Nothing to index
       logger.trace(`${this.prefix} No new blocks to index`);
       return;
     } else {
       logger.debug(
-        `${this.prefix} Indexing from block ${fromBlock} to ${toBlock}: ${numberOfBlocksToIndex} blocks`
+        `${this.prefix} Indexing from block ${fromBlock} to ${toBlockFinal}: ${totalBlocksToIndex} blocks`
       );
     }
 
+    // Process blocks in batches
+    while (fromBlock <= toBlockFinal) {
+      // Calculate toBlock for this batch
+      const toBlock =
+        fromBlock + MAX_BLOCKS_PER_BATCH - 1n > toBlockFinal
+          ? toBlockFinal
+          : fromBlock + MAX_BLOCKS_PER_BATCH - 1n;
+
+      // Process this batch of blocks
+      const producerState: TradeNotificationProducerState = {
+        lastBlock: toBlock.toString(),
+        lastBlockTimestamp: lastBlock.timestamp.toString(),
+        lastBlockHash: lastBlock.hash,
+      };
+      await this.processBlocks(fromBlock, toBlock, producerState);
+
+      // Move to next batch
+      fromBlock = toBlock + 1n;
+    }
+
+    // Check if during the process time there were some new blocks
+    const newLastBlock = await client.getBlock();
+    if (newLastBlock.number > toBlockFinal) {
+      logger.debug(
+        `${this.prefix} New blocks were indexed during the process: ${
+          lastBlock.number - toBlockFinal
+        } blocks`
+      );
+
+      // Recursive call to process the new blocks
+      return this.fetchAndSend();
```

Comment:
Question: couldn't running this recursively still blow up the memory due to the stack size?

Reply:
It should eventually catch up. Normally processing logs is fast, and block processing is ~10s (depend on chain).
What it could happen is that we take 300ms to process all blocks, and we are about to get a new block. This ensures we eagerly catch up instead of waiting for next poll

Reply:
Rabbit agrees with me =P https://github.com/cowprotocol/bff/pull/139#discussion_r2093319140

It's not about speed, but about the stack size.
Anyway, seems to have worked.
Still, I liked their last suggestion.